DRAFT: 4.0 build system

patrickarlt opened the pull request:
This is a major update and modernization of the build system to resolve a whole slew of things:

* https://github.com/Esri/arcgis-rest-js/issues/908
* https://github.com/Esri/arcgis-rest-js/issues/907
* https://github.com/Esri/arcgis-rest-js/issues/906
* https://github.com/Esri/arcgis-rest-js/issues/785
* https://github.com/Esri/arcgis-rest-js/issues/788
* Removes support for custom `fetch` (i.e. `IRequestOptions.fetch`)

Currently the work is only done for `arcgis-rest-request`. This is because we need to add `.js` to EVERY file extension to #907 since the module spec requires it. The Node JS tests require this. Once this gets general approval I can go through and add `.js` everywhere else.

I'll try to summarize the changes here and with inline comments:

* This is ~50% though the process of (#908). There are still some issues:
   * `npm version --ws` Does not update the `dependencies` or `peerDependencies` in the packages like `lerna` does. We either have to make a custom script that does this OR use `lerna` just for this one command or switch to `standard-version` with some custom processing to also get the changelog tools. I'm open for discussion on this. 
   * None of the release process script that use `lerna` have been updated yet.
* The `prepare` lifecycle script has been removed from all packages. This should probably get re-added before this gets merged. Or a global `"prepare": "npm run build -ws"` in the root `package.json`.
* Test matrix expanded to the latest Node 12, 14 and , 16.
* Adds 2 new demos for ES Modules and Common JS modules in Node
* Removes support for custom `fetch` in `IRequestOptions`. This greatly simplifies just about everything (mostly getting TypeScript checks to pass) These tests have been commented out for now.
* `arcgis-rest-request` now has a dependency on `formdata-node` which is an isomorphic `FormData` ponyfill for Node and `@esri/arcigs-rest-request-fetch` which is basically `cross-fetch` without a browser ponyfill. This remove the requirement for Node users to provide their own `fetch` and `FormData` polyfills.
* To further avoid side effects and allow tree shaking `setDefaultRequestOptions` now sets `DEFAULT_ARCGIS_REQUEST_OPTIONS` on [`globalThis`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/globalThis). A new `getDefaultRequestOptions` fetches the defaults. This might avoid an edge case in which `let DEFAULT_ARCGIS_REQUEST_OPTIONS = {}` got tree shaken out. See https://sgom.es/posts/2020-06-15-everything-you-never-wanted-to-know-about-side-effects/
* Adds a `engine` property in `package.json` set to `>=12.16.0`, this is when conditional exports was unflagged https://nodejs.org/api/packages.html#packages_conditional_exports
* The build process now works like:
   * Runs TypeScript  with no output to check for errors
   * Complies to CommonJS `dist/cjs`
   * Compiles to JavaScript Modules `dist/esm`
   * Compiles to UMD `dist/umd`
   * Creates `package.json`s in `dist/cjs` and `dist/esm` to markt he correct module type for Node.
   * Runs the [Skypack package checker](https://github.com/skypackjs/package-check) to make sure the package.json looks good for the Skypack CDN which will resolve #788.

In order to merge this:

* [ ] Add `.js` extensions to all imports under `packages/*`
* [ ] Add `engines` to all `package.json`s to match `@esri/arcgis-rest-request`
* [ ] Update `main`, `module`, `types` and `exports` in all `package.json`s to match `@esri/arcgis-rest-request`
* [ ] Re-enable all node tests under https://github.com/Esri/arcgis-rest-js/compare/4.0-build-system?expand=1#diff-dc637d0444a30c991c7473abfaf0eda2d3aade2b6d9809e74a80abe4d2077316R3
* [ ] Fix any broken Node tests (unknown what might break)
* [ ] Add `"type": "commonjs"` or `"type": "module"` to everything in `demo/*` 
* [ ] Confirm that all demos work.
* [ ] Confirm all demos are `"private": true` in their `package.jsons`
* [ ] Remove commended out tests for dependency checks (no longer needed) and custom fetch (removed)
* [ ] Fix release process (https://github.com/Esri/arcgis-rest-js/issues/688). Current recommendation:
   * [ ] Switch to `standard-version` with a [custom formatter](https://github.com/conventional-changelog/standard-version#can-i-use-standard-version-for-additional-metadata-files-languages-or-version-files) which also bumps all our `dependencies`.,`devDependencies`, `peerDependencies` and `optionalDepenendies`, in all packages.
   * [ ] Add a lint step that forces all commits to be conventional https://github.com/conventional-changelog/commitlint\
   * [ ] `npm publish --ws` runs `npm publish` in all workspaces
   * [ ] Create the zip archive and run `gh-release` a normal

patrickarlt commented:
Note that there is some risk in the dual CJS/EMS approach, mainly:

> When an application is using a package that provides both CommonJS and ES module sources, there is a risk of certain bugs if both versions of the package get loaded. This potential comes from the fact that the pkgInstance created by const pkgInstance = require('pkg') is not the same as the pkgInstance created by import pkgInstance from 'pkg' (or an alternative main path like 'pkg/module'). This is the “dual package hazard,” where two versions of the same package can be loaded within the same runtime environment. While it is unlikely that an application or package would intentionally load both versions directly, it is common for an application to load one version while a dependency of the application loads the other version. This hazard can happen because Node.js supports intermixing CommonJS and ES modules, and can lead to unexpected behavior.
>
> https://nodejs.org/api/packages.html#packages_dual_package_hazard

There are a few mitigating approaches to this mainly [Approach #2: Isolate state](https://nodejs.org/api/packages.html#packages_approach_2_isolate_state)

> This can be done if both the CommonJS and ES module versions of the package are equivalent, for example because one is the transpiled output of the other; and the package’s management of state is carefully isolated (or the package is stateless).

In general though I think we are OK since the entire library is stateless anyway. The only state we MIGHT use is `DEFAULT_ARCGIS_REQUEST_OPTIONS` which now exists on the global object if you call `setDefaultRequestOptions()` or state that the application is supposed to instantiate like `UserSesssion`

patrickarlt commented:
Additional consideration is this comment in the `node-fetch` example of the `formdata-node` repo:

> Note that `node-fetch` >= 3.x have builtin support for spec-compliant FormData, sou you'll only need the `form-data-encoder` if you use `node-fetch` <= 2.x.

Since we use `node-fetch@2.6.1` we might run into this although I question if "spec-compliant" is really necessary for our various use cases. I would REALLY appreciate someone else testing this with stuff like `addResources` or `addAttchment` with a practical tests case in Node since I am really not familiar with them. If we find issues I think we can work around this.

noahmulfinger commented on package.json:
```diff
@@ -215,5 +208,19 @@
     "packages/**/dist"
+  ],
+  "workspaces": [
+    "./packages/arcgis-rest-types",
+    "./packages/arcgis-rest-request",
+    "./packages/arcgis-rest-auth",
+    "./packages/arcgis-rest-demographics",
+    "./packages/arcgis-rest-feature-layer",
+    "./packages/arcgis-rest-geocoding",
+    "./packages/arcgis-rest-portal",
+    "./packages/arcgis-rest-request",
+    "./packages/arcgis-rest-routing",
+    "./packages/arcgis-rest-service-admin",
+    "./packages/arcgis-rest-fetch",
```
Can we use a wildcard here?
```suggestion
    "./packages/*",
```

patrickarlt replied:
@noahmulfinger you actually cannot because you need to compile things in dependency order. If you try to build `./packages/arcgis-rest-routing` before `./packages/arcgis-rest-request` you might get an error.

noahmulfinger replied:
@patrickarlt Ah okay. Well in that case, line 220 can be removed because its a duplicated entry of `arcgis-rest-request`

patrickarlt replied:
If we switch to https://www.npmjs.com/package/ultra-runner (#741) we can use the `--build` option but OOTB `npm run build --ws` (run `build`) in all workspaces might fail by default.

noahmulfinger commented on packages/arcgis-rest-feature-layer/src/package.json:
```diff
@@ -0,0 +1 @@
+{"type":"module"}
```
Is this file needed? I don't see any other packages with this `src/package.json` file. There's also a `test/package.json` file.

patrickarlt replied:
@noahmulfinger yes I should remove this.

noahmulfinger commented on packages/arcgis-rest-request/package.json:
```diff
@@ -2,28 +2,41 @@
+    "prebuild": "tsc --noEmit",
+    "build": "npm run build:node && npm run build:esm && npm run build:umd",
+    "postbuild": "node ../../support/create-dist-package-jsons.js && npx package-check",
+
```
empty line
```suggestion
```

noahmulfinger commented on packages/arcgis-rest-request/src/utils/encode-form-data.ts:
```diff
@@ -1,23 +1,23 @@
-import { encodeQueryString } from "./encode-query-string";
+import { processParams, requiresFormData } from "./process-params.js";
+import { encodeQueryString } from "./encode-query-string.js";
+import { FormData } from "formdata-node";
```
Will this package be used instead of the browser's FormData even when running the browser?

patrickarlt replied:
Yes but if you look `formdata-node` is isomorphic and just returns the browsers `window.FormData`. See https://github.com/octet-stream/form-data/issues/38

noahmulfinger reviewed (commented):
@patrickarlt Looks great! I had a few questions and comments below.
- Will we need a way to enforce `.js` endings for imports? I imagine it will be easy to forget to do this, especially since the files in the repo aren't `.js` files.
- Are there any specific things I can run to test things out right now? I ran the new common-js and es-modules demos and they worked fine. I tried to the `build` command and lots of the demos fail to build (I'm assuming due to WIP).

tomwayson commented on umd-base-profile.js:
```diff
@@ -68,14 +71,19 @@ export default {
-    commonjs()
-  ]
+    nodeResolve({
+      browser: true, // prefer `browser` fields in package.json or package.json exports
```
Unless you are planning to bring them back our packages scrapped the `browser` field for `unpkg` a while back. Or is this for our dependencies? If the latter, maybe update this comment to that effect.

patrickarlt replied:
We have to bring back the `browser` #906. Currently the browser field is the only way to get different behavior between Node and Browsers. We use `browser: true` when we make OUR UMD bundle to bundle `formdata-node` into the UMD bundle. 

When users bundle their app with `@esri/arcgis-rest-*` as a dependency they will need to bundle `formdata-node` which will require `browser: true` for `@rollup/plugin-node-resolve`.

The `browser` field is also used by Parcel v1 although Parcel v2 switches to prefer `module`.

I think there MIGHT be a Rollup bug where they go `main` and skip `module` which is weird. I'll run some other tests with Vite, Parcel, Snowpack and WMR tomorrow to make sure we have good coverage.

tomwayson replied:
I don't have strong feelings about `package.browser` in and of itself. The reason we removed it before is b/c we were doing it wrong (#420). If you have reason to bring it back, that's fine.

> When users bundle their app with @esri/arcgis-rest-* as a dependency they will need to bundle formdata-node which will require browser: true for @rollup/plugin-node-resolve.

I'm fine w/ making it a requirement to hav it in all targets if [these numbers](https://bundlephobia.com/package/formdata-node@4.0.1) really are true:

![image](https://user-images.githubusercontent.com/662944/132691560-9f4308c8-74e2-4482-8393-6ae9a4829c58.png)

I'm still unclear on why we need any special behavior for that package, it seems like you have it set up as a normal dependency that we `import` from. It seems to me that application bundlers would just include it in w/o any special configuration. Is this just a rollup issue? Or a UMD bundling issue. If it's the latter, as I've said before, I'm in favor of dropping the UMD builds (which allows us to get rid of rollup), especially if they require us to change how we configure our packages.

patrickarlt replied:
@tomwayson I ended up finally figuring this out `formdata-node` defined the Node ESM module in their `package.json` `module` field so Rollup kept trying to bundle the node version by default. I ended up creating `@esri/arcgis-rest-form-data` which works around this so now the only requirement for this to work in Rollup is `@rollup/plugin-node-resolve` with default options 🙌 .

As for `browser` I read through all of https://github.com/mjackson/unpkg/issues/63 until I found this smoking gun

> This will fix a frequent issue webpack users experiencing consuming D3 as ES modules: webpack’s resolve.mainFields defaults to ["browser", "module", "main"] when the target is web, so by default webpack users get the pre-built UMD bundle rather than the ES modules.
> 
> https://github.com/mjackson/unpkg/issues/63#issuecomment-326746759

Back in 2017 this would have been a problem and broken tree shaking.  But now Webpack will use what we define in `exports` instead of `browser` which will be `exports > import` which is EMS.

But this doesn't matter anyway since we would just be setting `browser` to `dist/cjs/index.js` which is **_identical_** `main`. The only modules that need `browser` fields are `@esri/arcgis-rest-fetch` and  `@esri/arcgis-rest-form-data` which DO define different behavior for browsers/node.

I'm pretty sure this means we still leave the `browser` field out of all the packages except `@esri/arcgis-rest-fetch` and  `@esri/arcgis-rest-form-data` which don't go on the CDN anyway.

tomwayson reviewed (commented):
Heroic work @patrickarlt 

I'm on board w/ all the high level changes you're proposing, other than the comments I left around `DEFAULT_ARCGIS_REQUEST_OPTIONS`. If we do end up keeping that as a global as you've done in this PR we should probably change the name to something like `__ARCGIS_REST_REQUEST_DEFAULT_OPTIONS__` or something.

Once everyone's on board w/ the high level changes, seems like there are a lot of changes in your housekeeping list that could be farmed out amongst the team, maybe on a package by package basis.

patrickarlt commented:
> Will we need a way to enforce .js endings for imports? I imagine it will be easy to forget to do this, especially since the files in the repo aren't .js files.

@noahmulfinger maybe we can do https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/file-extension-in-import.md. I might try to add this JUST to run it with `--fix` to migrate all the files to `.js` 😄 

> Are there any specific things I can run to test things out right now? I ran the new common-js and es-modules demos and they worked fine. I tried to the build command and lots of the demos fail to build (I'm assuming due to WIP).

@noahmulfinger The `tree-shaking-webpack` and `tree-shaking-rollup` demos should also work now. I did run into https://github.com/octet-stream/form-data/issues/44 but it has a fairly simple workaround even if nothing changes in `formdata-node`. Which will need to be documented as a caveat for Rollup.

noahmulfinger commented on umd-base-profile.js:
```diff
@@ -55,7 +58,7 @@ const globals = packageNames.reduce((globals, p) => {
   return globals;
 }, {});
-
+console.log({ packageNames, globals });
```
```suggestion
```

noahmulfinger reviewed (commented):
@patrickarlt Adding the linter rule sounds good. Also, I just pushed up a small update to package-lock.json after doing a local install.

patrickarlt commented:
@noahmulfinger so there IS a linter rule 🙌 but it only supports bare imports without extensions and it thinks the extension should be `.ts`and not `.js` :( https://github.com/import-js/eslint-plugin-import/issues/2111#issuecomment-915663474

patrickarlt commented on packages/arcgis-rest-request/src/request.ts:
```diff
@@ -368,9 +375,17 @@ export function request(
+       * configuration of fetch-mock in tests.
+       */
+
+      /* istanbul ignore next coverage is based on browser code and we don't test for the absence of global fetch so we can skip the else here. */
```
`istanbul ignore else` is only for `if/else` apparently. If you want to ignore coverage on a ternary you have to just ignore the whole thing. If you want to try it out you can change this to  `istanbul ignore else` and run `npm run test:chrome` and it should show some missing branch coverage.

patrickarlt commented on packages/arcgis-rest-request/src/request.ts:
```diff
@@ -35,15 +29,26 @@ let DEFAULT_ARCGIS_REQUEST_OPTIONS: IRequestOptions = {
     );
   }
-  DEFAULT_ARCGIS_REQUEST_OPTIONS = options;
+  (globalThis as any).DEFAULT_ARCGIS_REQUEST_OPTIONS = options;
+}
+
+export function getDefaultRequestOptions() {
+  return (
+    (globalThis as any).DEFAULT_ARCGIS_REQUEST_OPTIONS || {
+      httpMethod: "POST",
+      params: {
+        f: "json",
+      },
+    }
+  );
```
We came to the same conclusion here because I'm note sure either after reading that same article. I think the safer bet is to just use `setDefaultRequestOptions()`/`getDefaultRequestOptions()` and the global. 

`globalThis` exists in all our target environments so we can safely store any overrides AND it is shared across the whole application state which is ideal. I'm pretty sure this avoids all side effects because nothing happens when the module is loaded.

I'm also pretty sure that the old non-global approach will run into issues with the "dual package hazard" mentioned in https://github.com/Esri/arcgis-rest-js/pull/911#issuecomment-915357783 because `const DEFAULT_ARCGIS_REQUEST_OPTIONS` would be different if the application loaded both the CJS and ESM versions at the same time but now they share a single global.

tomwayson replied:
the non-global approach can definitely run into the the "dual package hazard", so I guess having the global is better.

patrickarlt commented:
It also appears like a lot of browser demos rely on built code in `node_modules` that has now moved to `../../node_modules` with the NPM workspaces move. I'll see if I can work out how to get these to work.

patrickarlt commented on packages/arcgis-rest-request/src/request.ts:
```diff
@@ -250,7 +256,7 @@ export function request(
   let originalAuthError: ArcGISAuthError = null;
 
-  const fetchOptions: RequestInit = {
+  const fetchOptions: any = {
```
@tomwayson reverting this back to `RequestInit` works just fine.

patrickarlt commented on packages/arcgis-rest-fetch/index.types.d.ts:
```diff
@@ -0,0 +1,14 @@
+/// <reference lib="dom" />
```
I removed this line. I originally copied it from `cross-fetch`

patrickarlt commented on packages/arcgis-rest-request/src/utils/encode-form-data.ts:
```diff
@@ -1,23 +1,23 @@
   const newParams = processParams(params);
   if (useFormData) {
-    const formData = new FormData();
+    const formData = new (FormData as any)();
```
This is also unnecessary now. The types from `formdata-node` just force the browser types onto it.

patrickarlt commented:
Since this seems to have met with general approval I'm going to keep working on the remaining items. We should probably have a conversation about if I need to fix the existing release process as a part of this PR or if we want to also rewrite the release tooling at 4.0 https://github.com/Esri/arcgis-rest-js/issues/688

patrickarlt commented:
Made lots of progress cleaning up edge cases today:

* Rollup works OOTB this was an issue in `formdata-node` that I worked around https://github.com/octet-stream/form-data/issues/44#issuecomment-916486724 by adding a new `@esri/arcgis-rest-form-data` package.
* An linter for enforcing file extensions is now in place. It does however incorrectly report that the files should end in `.ts` not `.js` https://github.com/import-js/eslint-plugin-import/issues/2111
* All release scripts were removed in prep for @dbouwman to migrate to `semantic-release`
* VS Code debugging now works in the Node and Chrome tests. To do this:
    * `npm run test:node:debug` - then run **Attach to `npm run test:node:debug`** in the VS Code debugger. You will be dropped in at a breakpoint at the start of the program and can run it from.
    * `npm run test:chrome:debug` -  then run **Attach to `npm run test:chrome:debug`** in the VS Code debugger. Start editing source or tests and the code should recompile and hit any breakpoints.
 * Added an ESM version of the `generate-sri-hashes.js` script for the SRI hashes for UNPKG. We will need to bundle this in each release.

patrickarlt commented:
Progress update:

* Merged `arcgis-rest-auth` with `arcgis-rest-request`. Tests are now run ONLY against `arcgis-rest-request` while I migrate other packages because a ton of stuff relied on the test helpers in `auth`.
* Because we define `exports` you can no longer import arbitrary stuff in node. This affects `@esri/arcgis-rest-auth/test/util`. Because it isn't defined in `exports` Node errors. I've moved this out of packages to a new location where we can share helpers across all tests.
* The `demos/oauth2-browser` should now work. This required:
   * Since NPM workspaces no longer install the packages under `demos/*/node_modules` I made a server script at `run-demo-server.js` that runs `browser-sync` in the current folder and maps all URLs starting with `@esri/arcgis-rest-*` to their respective packages. The `start` script just points to this shared script now.
* Created a shared util to get the `package.json` content of all `packages/*/package.json` files. This is used by the demo server script and the SRI hash script.

There is a MAJOR bug that I found which will break backward compat in some cases:

1. If you are using the UMD modules
2. As globals
3. And you include both `request` and `auth`
4. There is an infinite recursion error because `request` defines its exports on the global `arcgisRest` then `auth` defines all its exports on the `arcgisRest` global as GETTERS that return the `request` exports. So this means that `arcgisRest.UserSession` just looks like 

    ```
      Object.defineProperty(global.arcgisRest, 'UserSession', {
        enumerable: true,
        get: function () {
          return global.arcgisRest.UserSession;
        }
      });
    ```

   The way around this would probably be to author a special UMD for `auth` by  hand that just didn't define these if they existed which I think is reasonable.

patrickarlt commented:
Progress update:

* `@esri/arcgis-rest-routing` migrated
* `@esri/arcgis-rest-portal` migrated
* `@esri/arcgis-rest-service-admin` migrated

I've disabled 3 failing tests with `xit`:

```
Pending:

1) search Authenticated methods should add binary item data by id
  Temporarily disabled with xit


2) get-unique-service-name: does single check if unique
  Temporarily disabled with xit


3) get-unique-service-name: makes multiple calls if already taken
  Temporarily disabled with xit
```

The "get-unique-service-name" tests spy on the module exports which isn't allowed because ES module imports are immutable. This can be updated to work with `fetchMock`.

The "should add binary item data by id" test also appears to be failing. I'm not sure if this is a bug in `fetchMock`, `node-fetch` or `formdata-node`. It would be nice if someone could test that file handling (item upload, attachments, ect...) work in Node.

patrickarlt commented:
More progress:

* All packages have been updated to .js line ending
* All tests have been re-enabled.
* All test and lint pass with the exception of missing coverage due to the 3 ignored tests (see above)
* The `attachments` and `ago-node-cli` demos now work.

I did notice 2 issues with the demos:;

1. In `ago-node-cli` the search call was being made with a `searchForm` parameter. It look slike this was never updated when we broke it at v2 https://github.com/Esri/arcgis-rest-js/blob/master/CHANGELOG.md#esriarcgis-rest-portal-25
2. In `attachments` I cannot seem to configure a service to allow me to add attachments without auth. I have https://services9.arcgis.com/tPT6CGkx2VJKZshp/arcgis/rest/services/trails/FeatureServer/0 configured with public editing but I cannot get adding attachments to work without adding a token. I think this is a backend issue not a RES JS issue but I would appreciate someone looking  at it.

COV-GIS commented on package.json:
```diff
@@ -2,218 +2,167 @@
+  "contributors": [
+    "Anthony Lukach",
+    "Bill Reynolds <wreynolds@esri.com>",
+    "City of Veronia GIS",
```
@patrickarlt Vernonia is misspelled here and elsewhere.

patrickarlt replied:
Fixed in https://github.com/Esri/arcgis-rest-js/pull/911/commits/8c58d3d85daadff38c1e073ea285f52df764c007 @COV-GIS!

patrickarlt commented:
Updated the `batch-geocoder-node`, `express` and `oauth2-browser-retry` demos

patrickarlt commented:
All demos have been updated and are working. However  `demos/webmap-checker-sapper` probably needs to be rewritten since sapper is now sveltekit and has had a top-bottom reinventing.

We also now export `package.json` from the node `exports` to avoid https://github.com/uuidjs/uuid/issues/444#issuecomment-630441826

patrickarlt commented:
More additions/fixes

* Switched to the `import type`/`export type` format when using types from external packages. This makes bundles smaller and prevents some Rollup errors.
* Added a `test:compile` to confirm that everything compiles with TypeScript before running other tests.
* Added tooling to format `package.json`s on commit and as an npm command. This makes it WAY easier to compare changes to `package.json`s accross the repo.
* New `browser-es-modules` demo showing native ES modules in the browser with an `importmap`, polyfilled for FF and Safari. We can use this as  a base for future demos.
* New `snowpack` demo for Snowpack
* New `parcel` demo for Parcel
* New `vite` demo for Parcel
* We also now export `package.json` from the node `exports` to avoid https://github.com/uuidjs/uuid/issues/444#issuecomment-630441826
* We now inline all file content in the sourcemaps. Parcel complained about this.
* Rollup config has now moved to a single `rollup.js` config file that generates 4 builds:
    * `{packge}.umd.js` - UMD for people who need it.
    * `{packge}.umd.min.js` - Minified UMD for people who need it.
    * `{packge}.esm.js` - Native JS module with `@esri/arcgis-rest-fetch` and `@esri/arcgis-rest-form-data` bundled. Mostly for our own demos and people who want to use ArcGIS REST JS via script tags with import maps WITHOUT Skypack.
    * `{packge}.esm.min.js` - Above but minified

Commands in the `package.json`s accross the repo have been re-worked some use [`ultra-runner`](https://www.npmjs.com/package/ultra-runner) and some use [`npm-run-all`](https://www.npmjs.com/package/npm-run-all) to try to get the fastest build times:

In the root:

For all packages

* `npm run build` - builds all distributions of every package with `ultra`, inside each package builds are done in parallel with `npm-run-all`. Output is errors only.
* `npm run build:esm`, `npm run build:cjs`, `npm run build:bundled` - as as the above but only one of our target distributions.
* `npm run dev:esm`, `npm run dev:cjs`, `npm run dev:bundled` - runs the appropriate watch command in all packages.

For a specific package

* `npm run build -w @esri/arcgis-rest-request` - run all builds in a specific workspace
* `npm run dev -w @esri/arcgis-rest-request` - run all dev commands in a specific workspace
* `npm run build:esm -w @esri/arcgis-rest-request` - run the esm build in a specific workspace
* `npm run dev:esm -w @esri/arcgis-rest-request` - run the esm dev command in a specific workspace
* `npm run build:cjs -w @esri/arcgis-rest-request` - run the common js build in a specific workspace
* `npm run dev:cjs -w @esri/arcgis-rest-request` - run the common js dev command in a specific workspace
* `npm run build:bundled -w @esri/arcgis-rest-request` - run the rollup build in a specific workspace
* `npm run dev:bundled -w @esri/arcgis-rest-request` - run the rollup dev command in a specific workspace

patrickarlt commented:
PR summary for review:

#### Breaking Changes

* Remove support for IE 11 and Legacy Edge. We now compile for [ES 2017](https://caniuse.com/?search=ES%202017) as a target however we are not currently using any ES 2015 features. In General this means: 
    * Modern versions (last 2) of Chrome, Firefox, Edge, Safari on desktop and mobile devices. Many other versions/browsers will work as long as they support ES 2017, `fetch`, `FormData` and `Promise`.
    * Node 12.20+
* `fetch` is now bundled for Node JS via `node-fetch`.
* `FormData` is now bundled for Node JS via `formdata-node`. Since `formdata-node` supplies only a standards compliant `FormData` implementation steams and buffers are no longer supported as parameters. You will need to use the `File` or `Blob` classes from `formdata-node`, `@esri/arcgis-rest-form-data` or `@esri/arcgis-rest-request`.
* In `@esri/arcgis-rest-portal` the `addItemData()`, `createItem()`, `updateItem()`, `updateItemInfo()` and `createItemInFolder()` no longer accept a `data` parameter. Use `text` or `file` as documented. `data` was previously used in some tests.
* In `@esri/arcgis-rest-portal` the The `file` option in `addItemData()`, `createItem()`, `updateItem()`, `updateItemInfo()` and `createItemInFolder()` and  now needs to be a [`File`](https://developer.mozilla.org/en-US/docs/Web/API/File) or [`Blob`](https://developer.mozilla.org/en-US/docs/Web/API/Blob) object. See the `attachments-node` demo for how to work with the `File` and `Blob` classes in Node. In browsers this is likely already a `File` object from an `<input type="File">`.
* In `@esri/arcgis-rest-feature-layer` the The `attachment` option in `addAttachment()`, `updateAttachment()`, and  now needs to be a [`File`](https://developer.mozilla.org/en-US/docs/Web/API/File) or [`Blob`](https://developer.mozilla.org/en-US/docs/Web/API/Blob) object. See the `attachments-node` demo for how to work with the `File` and `Blob` classes in Node. In browsers this is likely already a `File` object from an `<input type="File">`.
* Support for supplying a custom `fetch` implementation in request options has been removed.
* `@esri/arcgis-request-auth` has been merged into `@esri/arcgis-rest-request`.

#### Additions/Changes

* Node JS users no longer need to supply `fetch` and `FormData` polyfills. Node JS is likely to add support for a native `fetch` and possibly `FormData` in a future version. Bundling our own polyfills helps prepare us for that future.
* Native support for ES modules in Node JS. Node supports ES modules unflagged since 12.17.0.See the [Modules: ECMAScript modules](https://nodejs.org/api/esm.html#esm_modules_ecmascript_modules) documentation.
* New bundled builds for supporting ES modules in browsers. 

#### Notes

* As of v4.0 we will also support ES modules via the Skypack CDN for developers that want to use ES modules in the browser without a build tool.
* Deno support (was likely possible in previous versions but should now be considered fully supported)
* React native support (was likely possible in previous versions but should now be considered fully supported)

noahmulfinger commented on package.json:
```diff
@@ -2,218 +2,174 @@
+    "./packages/arcgis-rest-feature-layer",
+    "./packages/arcgis-rest-geocoding",
+    "./packages/arcgis-rest-portal",
+    "./packages/arcgis-rest-request",
```
This line is duplicated (162 and 168 are both `arcgis-rest-request`)

noahmulfinger reviewed (commented):
I went through some of the code and some demos and I have a few comments. This could all be work done separately on the v4 branch though if we want to just get this PR in first.

- Linting-wise, I notice there are a few files that have unused imports. Is the linter not catching that? 
    - `packages/arcgis-rest-portal/src/items/add.ts`
    - `packages/arcgis-rest-portal/src/items/helpers.ts`
    - `packages/arcgis-rest-request/src/federation-utils.ts`
- Some of the demos have package-lock.json files. Should those exist since we are using an npm workspace?
- The `browser-es-modules` demo has some errors, like "`request` is not defined". It also could use a little clean up because it looks like the JSON is being displayed as HTML.
- The `node-es-modules` and `node-common-js` demos are missing start scripts.
- The `snowpack` and `vite` demo ask users to run npm install, but that creates a `package-lock.json` file that isn't gitignored.

gavinr commented:
I ran `npm build` and `npm run test` in my environment, it looks good there. 

In terms of documentation updates, I'd like to see (And am happy to try to help contribute):
1. Update `README.md` (contains a reference to `lerna`, which this PR removes (I think), and is also probably generally outdated after this major update and probably needs a refresh.
    - I'm not sure we've ever had this, but it would be nice to add a "developer quick start" - like if I'm a developer not super familiar with the project, how do I get a quick dev environment going for a particular package (For example, making a change to the routing package)
2. `docs\HISTORY.md` - at minimum, "Architecture" needs update (or removal?) ... and maybe a general update to this file with new status.
3. `RELEASE.md` - If the plan is to update the release workflow after we merge this PR, maybe this will be updated then.
4. `docs\src\guides\node.md` needs updating, possibly other pages in the guide section as well.

patrickarlt commented on packages/arcgis-rest-auth/src/index.ts:
```diff
@@ -1,13 +1,30 @@
-export * from "./app-tokens";
-
-export * from "./validate-app-access";
+export {
```
@noahmulfinger this is the major breaking change from 3.x > 4.x we are mostly doing this to that we don't have to immediately rewrite all tutorials and sections in the developer guide.

patrickarlt commented on package.json:
```diff
@@ -2,218 +2,174 @@
+    ],
+    "*.ts": [
+      "prettier --write --parser typescript --tab-width 2 --use-tabs false",
+      "eslint --project tsconfig.json",
```
TSLint was deprecated in 2019 https://palantir.github.io/tslint/ and merged with eslint. I there are only a few tslint:disable comments all in the tests which I think are still respected by the parser.

patrickarlt commented:
I think this is ready to merge. I addressed https://github.com/Esri/arcgis-rest-js/pull/911#issuecomment-921911447 and https://github.com/Esri/arcgis-rest-js/pull/911#pullrequestreview-756630660 and responded to @noahmulfinger's comments in https://github.com/Esri/arcgis-rest-js/pull/911#discussion_r710463852 and https://github.com/Esri/arcgis-rest-js/pull/911#discussion_r710386622

tomwayson commented:
I’m sure it is. I will take a look over this before EOB today, but if you are tired of waiting for me, go for it.

From: Patrick Arlt ***@***.***>
Date: Friday, September 24, 2021 at 4:26 PM
To: Esri/arcgis-rest-js ***@***.***>
Cc: Tom Wayson ***@***.***>, Mention ***@***.***>
Subject: Re: [Esri/arcgis-rest-js] 4.0 build system (#911)

I think this is ready to merge. I addressed #911 (comment) [github.com]<https://urldefense.com/v3/__https:/github.com/Esri/arcgis-rest-js/pull/911*issuecomment-921911447__;Iw!!CKZwjTOV!ifACSNt-P3z2_5pG90ZR7_mPbILE6kdjekSGcwGckKd1zkBe5QT5aPpKcH2t$> and #911 (review) [github.com]<https://urldefense.com/v3/__https:/github.com/Esri/arcgis-rest-js/pull/911*pullrequestreview-756630660__;Iw!!CKZwjTOV!ifACSNt-P3z2_5pG90ZR7_mPbILE6kdjekSGcwGckKd1zkBe5QT5aCHloSDx$> and responded to @noahmulfinger [github.com]<https://urldefense.com/v3/__https:/github.com/noahmulfinger__;!!CKZwjTOV!ifACSNt-P3z2_5pG90ZR7_mPbILE6kdjekSGcwGckKd1zkBe5QT5aEFmZbYz$>'s comments in #911 (comment) [github.com]<https://urldefense.com/v3/__https:/github.com/Esri/arcgis-rest-js/pull/911*discussion_r710463852__;Iw!!CKZwjTOV!ifACSNt-P3z2_5pG90ZR7_mPbILE6kdjekSGcwGckKd1zkBe5QT5aDt3x5wU$> and #911 (comment) [github.com]<https://urldefense.com/v3/__https:/github.com/Esri/arcgis-rest-js/pull/911*discussion_r710386622__;Iw!!CKZwjTOV!ifACSNt-P3z2_5pG90ZR7_mPbILE6kdjekSGcwGckKd1zkBe5QT5aHbW0CWG$>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub [github.com]<https://urldefense.com/v3/__https:/github.com/Esri/arcgis-rest-js/pull/911*issuecomment-926968166__;Iw!!CKZwjTOV!ifACSNt-P3z2_5pG90ZR7_mPbILE6kdjekSGcwGckKd1zkBe5QT5aC95bO8b$>, or unsubscribe [github.com]<https://urldefense.com/v3/__https:/github.com/notifications/unsubscribe-auth/AAFB3IHKHM3XRRE5JTERL5DUDUCL3ANCNFSM5DTEPNSA__;!!CKZwjTOV!ifACSNt-P3z2_5pG90ZR7_mPbILE6kdjekSGcwGckKd1zkBe5QT5aADWbsme$>.
Triage notifications on the go with GitHub Mobile for iOS [apps.apple.com]<https://urldefense.com/v3/__https:/apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675__;!!CKZwjTOV!ifACSNt-P3z2_5pG90ZR7_mPbILE6kdjekSGcwGckKd1zkBe5QT5aCH2r7kK$> or Android [play.google.com]<https://urldefense.com/v3/__https:/play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign*3Dnotification-email*26utm_medium*3Demail*26utm_source*3Dgithub__;JSUlJSU!!CKZwjTOV!ifACSNt-P3z2_5pG90ZR7_mPbILE6kdjekSGcwGckKd1zkBe5QT5aPKMytDW$>.

tomwayson reviewed (approved):
Heroic work @patrickarlt, let's get this merged!
